api-put-object.go

```diff
@@ -18,119 +18,73 @@ package minio
 
 import (
 	"bytes"
+	"context"
 	"fmt"
 	"io"
-	"os"
-	"reflect"
-	"runtime"
+	"net/http"
 	"runtime/debug"
 	"sort"
 	"strings"
 
+	"github.com/minio/minio-go/pkg/encrypt"
 	"github.com/minio/minio-go/pkg/s3utils"
 )
 
-// toInt - converts go value to its integer representation based
-// on the value kind if it is an integer.
-func toInt(value reflect.Value) (size int64) {
-	size = -1
-	if value.IsValid() {
-		switch value.Kind() {
-		case reflect.Int:
-			fallthrough
-		case reflect.Int8:
-			fallthrough
-		case reflect.Int16:
-			fallthrough
-		case reflect.Int32:
-			fallthrough
-		case reflect.Int64:
-			size = value.Int()
-		}
-	}
-	return size
+// PutObjectOptions represents options specified by user for PutObject call
+type PutObjectOptions struct {
+	UserMetadata       map[string]string
+	Progress           io.Reader
+	ContentType        string
+	ContentEncoding    string
+	ContentDisposition string
+	CacheControl       string
+	EncryptMaterials   encrypt.Materials
+	NumThreads         uint
 }
 
-// getReaderSize - Determine the size of Reader if available.
-func getReaderSize(reader io.Reader) (size int64, err error) {
-	size = -1
-	if reader == nil {
-		return -1, nil
+// getNumThreads - gets the number of threads to be used in the multipart
+// put object operation
+func (opts *PutObjectOptions) getNumThreads() (numThreads int) {
+	if opts.NumThreads > 0 {
+		numThreads = int(opts.NumThreads)
+	} else {
+		numThreads = totalWorkers
 	}
-	// Verify if there is a method by name 'Size'.
-	sizeFn := reflect.ValueOf(reader).MethodByName("Size")
-	// Verify if there is a method by name 'Len'.
-	lenFn := reflect.ValueOf(reader).MethodByName("Len")
-	if sizeFn.IsValid() {
-		if sizeFn.Kind() == reflect.Func {
-			// Call the 'Size' function and save its return value.
-			result := sizeFn.Call([]reflect.Value{})
-			if len(result) == 1 {
-				size = toInt(result[0])
-			}
-		}
-	} else if lenFn.IsValid() {
-		if lenFn.Kind() == reflect.Func {
-			// Call the 'Len' function and save its return value.
-			result := lenFn.Call([]reflect.Value{})
-			if len(result) == 1 {
-				size = toInt(result[0])
-			}
-		}
+	return
+}
+
+// getMetaData - constructs the headers from metadata entered by user in
+// PutObjectOptions struct
+func (opts *PutObjectOptions) getMetadata() (headers map[string][]string) {
```

Comment:
We can make the return value to be more typed i.e `getMetadataHeaders() (headers http.Header)` instead.

Reply:
done.